### PR TITLE
Adding implementation of IEquatable to Memory and ReadOnlyMemory

### LINKED ISSUE #32905: Should Memory<T> implement IEquatable<T>?
In the [.NET Standard](https://github.com/dotnet/standard/pull/909#discussion_r226390377) review of `Span<T>` @marek-safar wrote:

> Why it does not implement `IEquatable<T>` ?
> ...
> Well, if it's hard why do you have `public bool Equals(System.Memory<T> other)` ? I think it'd make sense to have the interface here if the implementation is there too.

It seems odd to providing a strongly typed `Equals` method but not implementing `IEquatable<T>`.

@KrzysztofCwalina @stephentoub @GrabYourPitchforks @jkotas

### PR DESCRIPTION
**Pull request by ostorc**
Adding implementation of IEquatable to Memory and ReadOnlyMemory. Resolves #32905

### CODE REVIEW

**Comment by stephentoub** [comment]
In addition to updating the refs, could you also add some tests that validate equality via the interfaces?

Thanks for working on this.

File: src/Common/src/CoreLib/System/Memory.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace System
     /// </summary>
     [DebuggerTypeProxy(typeof(MemoryDebugView<>))]
     [DebuggerDisplay("{ToString(),raw}")]
-    public readonly struct Memory<T>
+    public readonly struct Memory<T> : IEquatable<Memory<T>>
```
Comment (ahsonkhan) [inline src/Common/src/CoreLib/System/Memory.cs]: Changes to files in `CoreLib` need to made in the coreclr repo: https://github.com/dotnet/coreclr. Once its checked in, the changes will be mirrored here automatically. This PR (in corefx) should only update the `System.Runtime` reference assembly and add some tests.

1) Open up a PR in coreclr with the changes to the implementation:
https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Memory.cs
https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/ReadOnlyMemory.cs
2) Update this PR to change the type definition in the contract and add tests.
3) Wait for the mirror PR to merge.
4) Merge this change in with the PR which auto-updates the coreclr dependency.

This is what's causing the CI to fail:
```text
Build FAILED.

D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19121.4\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19121.4\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.ReadOnlyMemory<T>' does not implement interface 'System.IEquatable<System.ReadOnlyMemory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19121.4\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Runtime\netcoreapp-Windows_NT-Debug\System.Runtime.dll' [D:\a\1\s\src\System.Runtime\src\System.Runtime.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19121.4\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.Memory<T>' does not implement interface 'System.IEquatable<System.Memory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Memory\src\System.Memory.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19121.4\build\Microsoft.DotNet.ApiCompat.targets(72,5): error : CannotRemoveBaseTypeOrInterface : Type 'System.ReadOnlyMemory<T>' does not implement interface 'System.IEquatable<System.ReadOnlyMemory<T>>' in the implementation but it does in the contract. [D:\a\1\s\src\System.Memory\src\System.Memory.csproj]
D:\a\1\s\.packages\microsoft.dotnet.apicompat\1.0.0-beta.19121.4\build\Microsoft.DotNet.ApiCompat.targets(85,5): error : ApiCompat failed for 'D:\a\1\s\artifacts\bin\System.Memory\netcoreapp-Windows_NT-Debug\System.Memory.dll' [D:\a\1\s\src\System.Memory\src\System.Memory.csproj]
    0 Warning(s)
    6 Error(s)

```

Let me know if you have any questions around that. Thanks.

**Comment by ahsonkhan** [comment]
@ostorc, were you able to make progress in the coreclr repo? Let me know if you need any help. Thanks!

**Comment by danmosemsft** [comment]
Need this for .NET Standard 2.1

**Comment by ahsonkhan** [comment]
@ostorc, let me know if you plan to update the PR and make the change in coreclr. Otherwise, I can help move it forward.

**Comment by karelz** [comment]
@ostorc ping?

**Pull request closed without merging**